proposals/memory64/Overview.md

```diff
@@ -2,12 +2,188 @@
 
 ## Overview
 
-TODO
+### Structure
+
+* The [limits][syntax limits] structure is changed to use `u64`
+  - `limits ::= {min u64, max u64?}`
+
+* The [memory type][syntax memtype] structure is extended to have an index type
+  - `memtype ::= limits valtype`
+
+* The [memarg][syntax memarg] immediate is changed to allow a 64-bit offset
+  - `memarg ::= {offset u64, align u32}`
+
+
+### Validation
+
+* [Memory page limits][valid limits] are extended for `i64` indexes
+  - ```
+    ⊦ limits : 2**16
+    ----------------
+    ⊦ limits i32 ok
+    ```
+  - ```
+    ⊦ limits : 2**48
+    ----------------
+    ⊦ limits i64 ok
+    ```
+
+* All [memory instructions][valid meminst] are changed to use the index type,
+  and the offset must also be in range of the index type
+  - t.load memarg
+    - ```
+      C.mems[0] = limits it   2**memarg.align <= |t|/8   memarg.offset < 2**|it|
+      --------------------------------------------------------------------------
+                          C ⊦ t.load memarg : [it] → [t]
+      ```
+  - t.loadN_sx memarg
+    - ```
+      C.mems[0] = limits it   2**memarg.align <= N/8   memarg.offset < 2**|it|
+      ------------------------------------------------------------------------
+                        C ⊦ t.loadN_sx memarg : [it] → [t]
+      ```
+  - t.store memarg
+    - ```
+      C.mems[0] = limits it   2**memarg.align <= |t|/8   memarg.offset < 2**|it|
+      --------------------------------------------------------------------------
+                         C ⊦ t.store memarg : [it t] → []
+      ```
+  - t.storeN_sx memarg
+    - ```
+      C.mems[0] = limits it   2**memarg.align <= N/8   memarg.offset < 2**|it|
+      ------------------------------------------------------------------------
+                       C ⊦ t.storeN_sx memarg : [it t] → []
+      ```
+  - memory.size
+    - ```
+         C.mems[0] = limits it
+      ---------------------------
+      C ⊦ memory.size : [] → [it]
+      ```
+  - memory.grow
+    - ```
+          C.mems[0] = limits it
+      -----------------------------
+      C ⊦ memory.grow : [it] → [it]
+      ```
+  - (and similar for memory instructions from other proposals)
+
+* [Data segment validation][valid data] uses the index type
+  - ```
+    C.mems[0] = limits it   C ⊦ expr: [it]   C ⊦ expr const
+    -------------------------------------------------------
+          C ⊦ {data x, offset expr, init b*} ok
+    ```
+
+
+### Execution
+
+* [Memory instances][exec mem] are extended to have 64-bit vectors and a `u64`
+  max size
+  - `meminst ::= { data vec64(byte), max u64? }`
+
+* [Memory instructions][exec meminst] use the index type instead of `i32`
+  - `t.load memarg`
+  - `t.loadN_sx  memarg`
+  - `t.store memarg`
+  - `t.storeN memarg`
+  - `memory.size`
+  - `memory.grow`
+  - (spec text omitted)
+
+* [memory.grow][exec memgrow] is changed to check for a size greater than
+  2<sup>64</sup> - 1, and to return 2<sup>64</sup>-1 when `memory.grow` fails.
```

Comment:
The value would now depend on the index type, wouldn't it?

Reply:
Yep, fixed.